core/stats_handler.go

```diff
@@ -0,0 +1,203 @@
+			if pl.SaveReply.TransactionError == "" {
+				writes = 1
+			}
+		case *tpb.DeleteReply:
+			if pl.TransactionError == "" {
+				writes = 1
+			}
+		case *tpb.WriteTransactionReply_DeleteReply:
+			if pl.DeleteReply.TransactionError == "" {
+				writes = 1
+			}
+		case *tpb.HasReply:
+			if pl.TransactionError == "" {
+				reads = 1
+			}
+		case *tpb.ReadTransactionReply_HasReply:
+			if pl.HasReply.TransactionError == "" {
+				reads = 1
+			}
+		case *tpb.WriteTransactionReply_HasReply:
+			if pl.HasReply.TransactionError == "" {
+				reads = 1
+			}
+		case *tpb.FindReply:
+			if pl.Instances != nil {
+				reads = int64(len(pl.Instances))
+			}
+		case *tpb.ReadTransactionReply_FindReply:
+			if pl.FindReply.Instances != nil {
+				reads = int64(len(pl.FindReply.Instances))
+			}
+		case *tpb.WriteTransactionReply_FindReply:
+			if pl.FindReply.Instances != nil {
+				reads = int64(len(pl.FindReply.Instances))
+			}
+		case *tpb.FindByIDReply:
+			if pl.TransactionError == "" {
+				reads = 1
+			}
+		case *tpb.ReadTransactionReply_FindByIDReply:
+			if pl.FindByIDReply.TransactionError == "" {
+				reads = 1
+			}
+		case *tpb.WriteTransactionReply_FindByIDReply:
+			if pl.FindByIDReply.TransactionError == "" {
+				reads = 1
+			}
+		case *tpb.ListenReply:
+			if pl.Instance != nil {
+				reads = 1
+			}
+		}
+		ctx = handleStats(ctx, egress, reads, writes)
+
+	case *stats.End:
+		// Record usage
+		rs := getStats(ctx)
+		if rs == nil {
+			return
+		}
+		go func() {
+			ctx, cancel := context.WithTimeout(context.Background(), statsTimeout)
+			defer cancel()
+			// @todo: Consolidate these to one IncStats call.
+			if rs.egress > 0 {
+				if _, err := h.t.bc.IncNetworkEgress(ctx, rs.customerID, rs.egress); err != nil {
+					log.Errorf("stats: inc network egress: %v", err)
+				}
+			}
+			if rs.reads > 0 {
+				if _, err := h.t.bc.IncInstanceReads(ctx, rs.customerID, rs.reads); err != nil {
+					log.Errorf("stats: inc instance reads: %v", err)
+				}
+			}
+			if rs.writes > 0 {
+				if _, err := h.t.bc.IncInstanceWrites(ctx, rs.customerID, rs.writes); err != nil {
+					log.Errorf("stats: inc instance writes: %v", err)
+				}
+			}
+		}()
+	}
+}
+
+type statsCtxKey string
+
+type requestStats struct {
+	customerID string
+	egress     int64
+	reads      int64
+	writes     int64
+}
+
+func (h *StatsHandler) TagRPC(ctx context.Context, info *stats.RPCTagInfo) context.Context {
+	for _, ignored := range authIgnoredMethods {
+		if info.FullMethodName == ignored {
+			return ctx
+		}
+	}
+	ctx, _ = h.t.newAuthCtx(ctx, info.FullMethodName, false)
+	var account *mdb.Account
+	if org, ok := mdb.OrgFromContext(ctx); ok {
+		account = org
+	} else if dev, ok := mdb.DevFromContext(ctx); ok {
+		account = dev
+	}
+	// @todo: Account for users after User -> Account migration
+	if account == nil || account.CustomerID == "" {
+		return ctx
+	}
+	return context.WithValue(ctx, statsCtxKey("requestStats"), &requestStats{
```

Comment:
Adds a wrapper to the context for tracking stats for the lifetime of the request (chunks from streams, etc.)